### PR TITLE
Search widget

### LINKED ISSUE #2739: For ID based dashboard filters, provide a search widget for the target record's name
When a dashboard creator wires up a foreign key ID as the dimension to be filtered on, we current ask people to input the actual ID.

This is often not widely known and it would be much more useful to search for `Acme Co` instead of entering the company ID.

### PR DESCRIPTION
**Pull request by tlrobinson**
TODO:

* [x] case insensitive search (needs backend support #2370)
* [x] control over which fields/tables this is enabled for (may want to opt-in if we stick with typeahead)
* [ ] ideally we'd implement query cancelation in the backend/db #2399 (everything necessary on the frontend is implemented)
* [x] currently only shows the committed entity name if it was previously cached by a queries, i.e. not when you refresh the browser
* [ ] better caching (currently it doesn't re-run queries if it can just further filter on the client but no other caching)
* [ ] min number of characters before searching?
* [x] loading state?
* [x] conflating field values and remapping causes search widget to switch to a select widget when a remapping is loaded. store them separately (see `ADD_REMAPPINGS` action/reducer in `metabase/redux/metadata` and `remapping` in `metabase/selectors/metadata`)
* [x] Fix issue with parameter formatting not showing the remapped value
  * I believe this is because we use `field.parameterSearchField()` to determine which field to search, but then use `field.remappedField()` in the `RemappedValue` component. Maybe we shouldn't have `parameterSearchField` at all, but then we won't get entity name searching by default (only when remappings have been configured)

general:
* [x] don't clear search when clicking an option
* [x] don't clear search when losing focus
* [x] load remapped values when first mounting `FieldValuesWidget` (since they may not be loaded yet and we don't use `autoLoad`ing `RemappedValue` in search)
* [x] automatically load field values if `field.has_field_values === "list"`

filters:
* [x] allow clicking "add filter" when text has been entered
* [x] fix contains, does not contain, starts with, ends with
* [x] make sure we don't show values for `BETWEEN` filters, etc

sql parameters:
* [x] fix js errors
* [x] actually support FieldValuesWidget in SQL parameters

parameters:
* [x] multiple values support in parameters
* [x] merge category picker in parameters
* [x] polish popover

backend
* [x] @camsaul: hydrate `has_field_values` in `/api/database/:id/metadata` and  `/api/field/:id` (currently `null`)
* [x] @camsaul: don't hydrate `values` in all endpoints (except `/api/field/:id/values` of course)

design:
* [x] @kdoh @mazameli: decide on placeholder text for various permutations `(list values, search values, none) X (simple value, remapped pk, remapped fk)`

misc:
* [ ] @salsakran @camsaul @mazameli @kdoh decide what level of support we want in public/embedded questions/dashboards
   * would require additional endpoints and frontend logic, etc

Resolves #2739

### CODE REVIEW

**Comment by tlrobinson** [comment]
@metabase/core-developers How do we want to proceed with this?

**Comment by tlrobinson** [comment]
FYI I've merged the latest from `master` into this branch in case anyone wants to take a look. Currently seems to work best on dashboard parameters.

**Comment by mazameli** [comment]
I put this in Slack, but I'll put it in here for posterity:

- seems good to me that you can enter either an ID or entity name
- the search being case sensitive really threw me off
- this was compounded by the fact that there doesn’t seem to be a “no results” search state yet
- I think we should show the popup with at least a loading spinner right away so you know something’s happening
- kind of a bummer that, in cases where a user name is stored as a single string, you can’t search by last name since the search starts at the beginning of the string and doesn’t search anywhere else
- LOVE that i can use the arrow keys to cycle through possible options. I wish this was possible with more of our popover menus.
- now it feels a little weird that if I start typing an ID number, there isn’t any searching
- The placeholder text, `enter a value`, feels wrong now. Maybe, `search by name or ID`? I know it’s not always going to be a person’s name, obviously, but i think that “name” still works in cases of things like product name, venue name, etc.
- Feels odd that if I correctly type out the whole entity name, but hit enter instead of selecting the menu item from the popover, the filter doesn’t work correctly. I have to select it from the menu or type in the _ID_ number

**Comment by mazameli** [comment]
I just spun this up to see its status because @tlrobinson mentioned in one of our planning meetings that he wanted some eyes on it.  

You're probably already aware of this behavior, but the main problem I'm encountering is that when I start typing in the search box, after a few key presses, the input box suddenly turns into a collapsed checkbox while I'm typing:

screen shot 2017-08-18 at 3 52 43 pm

Opening the checkbox shows values that correspond to what I had started typing, but there's no way to clear out or change my filter string (unless I refresh the page):

screen shot 2017-08-18 at 3 52 46 pm

The size of the filter widget also doesn't seem to be large enough for the default text:

screen shot 2017-08-18 at 3 52 26 pm

Let me know if there's anything else I can test out, Tom.

**Comment by tlrobinson** [comment]
FYI I've `master` into this. It seems to still work.

**Comment by goodfella1408** [comment]
Is metabase planning to release this anytime soon .If I understand correctly with this change 
1)Filter values will be searchable
2)Filter values will be served directly from the DB rather than the metabase cache - which is a more feasible approach because different users might be authorized to see different values in the filter drop down for eg in a user drop down head of Asia/Pacific must see only users from Asia/Pacific rather than those from Eurpoe

**Comment by tlrobinson** [comment]
@goodfella1408 We are actively working on it. No promises on the timeframe though :)

**Comment by tlrobinson** [comment]
I've tried out two different approaches to get remapped values to show up in the filter/parameter widgets, i.e. when you filter by user ID = 1 in our sample dataset we want to show "Adelia Eichmann" in the widget:

screenshot 2018-01-29 17 51 19
screenshot 2018-01-29 17 50 47

Neither seems ideal but here they are:

1. [Add `type/PK` -> `type/Name` remappings](https://github.com/metabase/metabase/compare/dash-filter-search-widget...search-widget-option-1)
  * it just works using the code I had in this PR before, which worked for filters (using configured FK remappings) but not parameters (since those are based on PKs, plus I don't think we even load the `dimension` in the metadata in dashboards)
  * if we actually persist the remapping on the backend we'd need to make sure we don't actually remap PKs to names in queries.
  * also need to make sure the filter widget fields show the right things. I changed 
2. [Add a prop to override the `displayColumn` on the `<RemappedValue>` component](https://github.com/metabase/metabase/compare/dash-filter-search-widget...search-widget-option-1) (should probably be called `displayField`)
  * we should probably change the redux store to add another level of maps to field's `remappings` since you could technically have remappings for the same value to different fields (though i'm not sure that would happen in practice)

**Comment by tlrobinson** [comment]
@camsaul I've switched to your endpoints. Search is working but the remapping one isn't.

Previously it was doing this MBQL query, for example, which worked:

```json
{
  "database": 1,
  "type": "query",
  "query": {
    "source_table": 3,
    "filter": ["=", ["field-id", 23], 1],
    "fields": [["field-id", 23], ["field-id", 24]],
    "limit": 1
  }
}
```

Now it's doing this API call, which returns an empty body:

```
http://localhost:3000/api/field/23/remapping/24?value=1
```

I didn't see anything obviously wrong with your implementation though.

**Comment by camsaul** [comment]
@tlrobinson it looks like you're using a numeric `value` whereas we're keeping `value` as a string (since it comes back in that way from the URL param.) For this endpoint will the value always be a number (i.e. will it work if we always parse as integer?)

**Comment by tlrobinson** [comment]
@camsaul I think it needs to work for remapped strings as well. I imagine you'll have to look at the field type and parse if it's numeric.

**Comment by tlrobinson** [comment]
@camsaul I started on a fix in e2c5291, but it only parses integers.

**Comment by kdoh** [comment]
@tlrobinson 

Added a few ToDos based on playing a bit more with this. These were the biggest "Oh that's obvious" level things I noted. If someone feels froggy in regard to testing some of the possible remapping edge cases that'd be helpful.

- [ ] Search should be case sensitive (@tlrobinson)   
I think we want to let you type `k` if you're searching for Kyle . Bumped into this a time or two myself in testing

- [ ] Center and properly color the loading indicator. (@kdoh or @tom if it's convenient)
There's some light jumpiness between the loading and

- [ ] If the selection / typeahead box is focused and you hit enter to add a typeahead selection we should add the suggestion but not close the entire filter popover (@tlrobinson, I figure you're more familiar with the events)
I'm assuming this is partially due to some overlapping key press handlers. I'm guessing if the input isn't focused then we should still have enter commit the 


- [ ] Input box active border should probably match the color scheme(@kdoh, but ideally after @tom provides theme color hooks)

- [ ] Possible edge case / question: "Should we let you add the same value twice?"
This came up when I added a user by name and then typed their ID and hit enter. I think unless someone can convince me there's a good reason to let the same field value item be in the list twice we should just quickly flash the 

- [ ] Better initial instructions (placeholder text, etc). (@tom probably, since this might require smartly referencing and pluralizing the field)
We need a better first read to help users know that they can search for values if there aren't any locally. Part of this might just be placeholder text but I think for that to be good we'll want to have something akin to:

`Search for a ${field.display_name}`
"Search for a User"
"Search for an Order"
"Search states"

I think for it to immediately be clear that's the ideal scenario. If this works worse than I'm expecting in practice we can at least say "Search for a value"

**Comment by tlrobinson** [comment]
@kdoh Thanks.

@camsaul Can you handle making the search endpoint case insensitive?

**Comment by tlrobinson** [comment]
@kdoh @mazameli Regarding placeholder text, do we want to let people know they can either enter an ID or search by name?

Reminder, there are actually two dimensions (such an overloaded term...) to this:

1. no field values vs. "listed" field values vs. "searchable" field values
2. no remapping vs. remapping from an ID (PK or FK) to a something, generally a name (show both ID and remapped value?) vs. another type of remapping (generally only show remapped value?)

Therefore potentially 9 different cases. It might be a good idea for you guys to draw out a matrix and think through the differences, both for placeholder text and other behaviors (e.x. which value to show in the tokenfield "token" / filter widget / parameter widget)

**Comment by tlrobinson** [comment]
I've fixed a few of the issues @kdoh noted:

1. added some placeholder placeholders in the `render` method of `FieldValuesWidget` for @kdoh to play around with. Some of it can be kind of wonky ("Select a Tax", etc). I can add more logic to handle different cases.
2. centered the loader, made it a little bigger. feel free to tweak it.
3. prevented filter widget from closing when you press enter when the input is focused
4. parameterized the `TokenField` with a `color` prop

I also started integrating with parameters. There are some rough edges:

1. The UI/UX is pretty wonky, especially around replacing the parameter widget with a popover when you click it. Part of the problem is the FieldValuesWidget input is a different size from the ParameterValueWidget. It also feels weird to show it in an otherwise blank popover if there's no other content in the popover. Maybe we could shrink the `TokenField` to be the same size as the `ParameterValueWidget` and only show the popover behind if there is more than one row of items, or if there are options / search results to show?
2. It doesn't currently work for remapped fields, I believe because we're not hydrating the `dimensions` property  in the `/api/database/:id/metadata` endpoint.
3. We have that special way we handle field values so that embedded dashboard users can use them even if they don't have access to the raw metadata. I assume we want to use the same widget in that case?
  * Corollary: do we want to support (server-side) searching in those cases? We'll need new endpoint(s) for that.

**Comment by mazameli** [comment]
Currently I can't add a contains, does not contain, starts with, or ends with filter unless I select a full result, which isn't what I want to do:

screen shot 2018-02-06 at 10 23 37 am

The `add filter` button for numeric filters is also disabled until you hit return, which I only figured out accidentally:
screen shot 2018-02-06 at 10 28 54 am

If the input boxes loses focus, like if you click on `more options` or click on your desktop or another browser wind, any text that you'd entered in there previously gets cleared out:

screen shot 2018-02-06 at 10 30 09 am

I'm getting JS errors when trying to select a variable type or map a field filter:

```
settings.js:205 Error getting setting table.columns TypeError: Cannot read property 'cols' of undefined
    at Object.getDefault (Table.jsx:73)
    at getSetting (settings.js:196)
    at getSettings (settings.js:247)
    at selectors.js:212
    at index.js:76
    at index.js:36
    at index.js:90
    at index.js:36
    at Function.mapStateToProps [as mapToProps] (QueryHeader.jsx:41)
    at mapToPropsProxy (wrapMapToProps.js:43)
```
and
```
Uncaught (in promise) TypeError: Cannot read property 'length' of undefined
    at Function.getWidget (ParameterValueWidget.jsx:85)
    at ParameterValueWidget.render (ParameterValueWidget.jsx:127)
    at ReactCompositeComponent.js:793
    at measureLifeCyclePerf (ReactCompositeComponent.js:73)
    at ReactCompositeComponentWrapper._renderValidatedComponentWithoutOwnerOrContext (ReactCompositeComponent.js:792)
    at ReactCompositeComponentWrapper._renderValidatedComponent (ReactCompositeComponent.js:819)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:359)
    at ReactCompositeComponentWrapper.mountComponent (ReactCompositeComponent.js:255)
    at Object.mountComponent (ReactReconciler.js:43)
    at ReactCompositeComponentWrapper.performInitialMount (ReactCompositeComponent.js:368)
```

File: frontend/src/metabase/components/Value.jsx
Comment (salsakran) [inline frontend/src/metabase/components/Value.jsx]: I like how cleaner this reads

**Comment by mazameli** [comment]
If I type in a search term and then select one of the results, the search then gets cleared out and the rest of the results go away. This makes it really annoying to try and select several items that match the same term; I have to type the search again for each item I want to select.

screen shot 2018-02-06 at 11 11 00 am

**Comment by mazameli** [comment]
FYI, I'm blocked on testing how this works with SQL parameters due to the JS errors I mentioned above.

**Comment by mazameli** [comment]
I couldn't figure out how to get a dashboard category filter to allow me to select or search for more than one value. Once I selected one, I saw this:

screen shot 2018-02-06 at 11 17 45 am

But then when I tried to click into the text input area, it reverts back to this:

screen shot 2018-02-06 at 11 17 48 am

And I'm sure I'm stating the obvious here, but it looks like we need to do some polish with the popover itself:

screen shot 2018-02-06 at 11 16 27 am

**Comment by salsakran** [comment]
Is there meant to be a loading indicator when a search is going on? 

I type in things, there's a bit of a pause, and then the list below it is populated with possibilities

**Comment by tlrobinson** [comment]
> Is there meant to be a loading indicator when a search is going on?
>
> I type in things, there's a bit of a pause, and then the list below it is populated with possibilities

@salsakran Hmm, there's definitely a loading indicator, not sure why you're not seeing it:

screen shot 2018-02-07 at 3 12 45 pm

**Comment by mazameli** [comment]
Weird behavior when clicking on a search suggestion:

weird-filter-behavior

**Comment by mazameli** [comment]
I'm finding with dashboard params that when I try to clear out a value, it stays present in the widget:

clearing-out-value

**Comment by mazameli** [comment]
If I type in the full string for a field value, it disappears from the list of suggestions (this happens in the QB as well):

type-full-name

**Comment by tlrobinson** [comment]
@mazameli Thanks, I think I've addressed all of those issues (and added tests for most of them)

**Comment by mazameli** [comment]
As I type another term when I have a few selected tokens already, things start feeling cramped. I guess in this situation we'd want to let the typed input wrap down to a second line.

cramped

**Comment by mazameli** [comment]
Has https://github.com/metabase/metabase/pull/6864 (case-insensitive search) not been incorporated? I thought it had been working, but now I'm seeing this behavior:

case-sensitivity

**Comment by mazameli** [comment]
If I input text into a search box such that results are returned, and then clear it out, every time the search input box loses focus, it creates a Phantom Token (just like the new PT Anderson movie):

phantom-token

**Comment by camsaul** [comment]
@mazameli no, case insensitive was a different PR. I went ahead and merged it in just now so pull and it should be working

**Comment by mazameli** [comment]
I think something kind of funny is going on with unmapped foreign key field search. If I click on the accordion to show the FK-connected fields, and select the ID of that field, I can search through the IDs. But if I select the FK field itself in the list of fields in my source table, it doesn't show me results when I try to search for an ID:

fk-search

**Comment by mazameli** [comment]
@tlrobinson here's a stab at a placeholder text matrix. I noted some places where I wasn't 100% sure what you meant. WDYT?

screen shot 2018-02-12 at 5 27 20 pm

**Comment by tlrobinson** [comment]
@mazameli thanks.

> None (I assume you mean "no results"?)

"none" means we don't have a list of values and we don't allow searching, either because it's something that doesn't make sense to show as a list/search (non-enum numbers, descriptions, JSON, etc), or it was disabled for performance reasons. You just have to enter a value manually. But I can add the empty state you described as well.

> Not sure I understand. Can you remap a primary key?

We do implicitly remap `type/PK`s to `type/Name` fields in the parameter widget, but not filter widget.

> "Search for an ID, like [first ID in the list]"

I'm not sure we can easily do "like [first ID in the list]". I guess I could fire off a search with search string of "" and limit of 1. Is that really necessary? Would anyone be searching for an ID if they don't know the format of the ID?

> "Enter an ID number"

We actually don't let you search for a numeric ID number so this would never happen

> Do we ever just list all the IDs?

Probably not by default, but the admin could change the setting so we do


Also, regarding the "list" and "search" cases, in all cases we allow you to enter a value manually rather than selecting from the list/search results, so maybe that should be incorporated into the "list" and "search" cases? We do want to keep the placeholder text concise so it fits, though we could show a more verbose explanation in the space below the input (except for "list" since that's taken by the list of values right away.

It might be easier to continue this in a shared Google Sheet, assuming you have one or can copy your matrix to one

**Comment by mazameli** [comment]
> It might be easier to continue this in a shared Google Sheet, assuming you have one or can copy your matrix to one.

Yep, great feedback, and we'll continue sorting this out in a doc. Link forthcoming.

**Comment by mazameli** [comment]
As you use the arrow keys to move the highlight through a list of values, it should scroll the list, but right now it just goes offscreen:

arrow-scroll

**Comment by mazameli** [comment]
screen shot 2018-02-14 at 5 00 27 pm

**So for this bit, here's what I'm thinking below. I'm realizing I'm not 100% clear what "Disable" does — turns off search, or prevents this field from being used in filters? If the former, doesn't "List all values" implicitly turn off search?**

---

{heading} Filters
{subheading} Choose whether you want to allow searching through this field's values in a filter, if you want to force filters to always display a list of all the values, or if you want to turn search off for this field entirely.

{dropdown}
- Allow searching for values
- Always list all values [I added the "Always" because my understanding is that the "Allow searching for values" will still show a list of all values provided the number of values is less than 300.]
- Disable search

**Comment by tlrobinson** [comment]
@mazameli turns off search and listing of values (these 3 options correspond to the 3 rows in your placeholder text document)

**Comment by tlrobinson** [comment]
> I added the "Always" because my understanding is that the "Allow searching for values" will still show a list of all values provided the number of values is less than 300.

@mazameli That wasn't my interpretation. If the admin changes from "list" to "search" then we won't load the field values up front. The 300 values limit is just part of the criteria for the default for this setting.

@salsakran can you clarify ^

**Comment by mazameli** [comment]
>> I added the "Always" because my understanding is that the "Allow searching for values" will still show a list of all values provided the number of values is less than 300.

> @mazameli That wasn't my interpretation. If the admin changes from "list" to "search" then we won't load the field values up front. The 300 values limit is just part of the criteria for the default for this setting.

Ah, I see, that makes sense. Then how about we can change things like this:

```
{heading} Filtering on this field
{subheading} When this field is used in a filter, what should people use to enter the value they want to filter on?
- Search box
- A list of all values
- Plain input box
```

**Comment by mazameli** [comment]
Been playing with this a bit on a table that I know is slow, and I'm not exactly sure what to take away from the experience. If I try to search for a value in one of the fields in this table, it can take between 20 to 40 seconds to return results, which of course is slow. But the combination of me knowing ahead of time that the table has a whole ton of columns and rows, and performs slowly compared to other tables in general, coupled with the fact that now I actually can search for hard-to-remember values whereas I couldn't before — I _think_ that's leaving me with a net favorable impression.

One thing I'm having a hard time replicating is that sometimes when it takes a long time to return search results (like around 40 seconds), the loading indicator will go away but nothing will be displayed for about 3 seconds, making me think nothing was found; but then the results pop into place a second later.

**Comment by salsakran** [comment]
@mazameli how big is that table? And how many distinct values are in that column?

**Comment by salsakran** [comment]
```
> I added the "Always" because my understanding is that
> the "Allow searching for values" will still show a list of
> all values provided the number of values is less than 300.


@mazameli That wasn't my interpretation. 
If the admin changes from "list" to "search" then we won't 
load the field values up front. The 300 values limit is just
part of the criteria for the default for this setting.
```

Yup. My understanding matches @tlrobinson 's. There is nothing at all magical about 300 entries after this PR. We'll probably still use it as a cut-off for `hasFieldValues` (i.e. we'll cache and display these values in a list), but once a field is a "list" it stays a list. I expect to tweak this going forward and I've been trying to play with it on slower databases.

Actually, a good question to ask is -- how large a list should be displayed (vs searched for?)
Also, implicit in marking something as being listed is that it should have the FieldValues fetched for it at that point in time.

**Comment by salsakran** [comment]
@mazameli using this makes me want to significantly dial up the caching multiplier (from the automatic 10x the query execution) for these queries.

**Comment by mazameli** [comment]
> how big is that table? And how many distinct values are in that column?

@salsakran 161 fields, 2 million rows, and the field has just under 3 thousand distinct values. It's the `Events` table in the baseball DB, and the `Bat ID` field, in case you want to try yourself.

**Comment by salsakran** [comment]
one weird thing is that if I erase characters down to 0 left, I get a loading indicator.

**Comment by mazameli** [comment]
> one thing that's a bit odd to me is that we leave the text that was auto-completed after I select an item

Me too. My assumption was that this is the case so that if you want to select more than one item that matches your search term, you don't have to type the term in again. But I wonder if we can make it so that the search term is cleared out after selecting an item, but that the search results persist until you enter a new search term.

**Comment by tlrobinson** [comment]
> one thing that's a bit odd to me is that we leave the text that was auto-completed after I select an item. I.e.

This was requested earlier in the PR but I guess I didn't understand the text should be cleared but the results should stay the same. I'll fix it.

**Comment by tlrobinson** [comment]
> one weird thing is that if I erase characters down to 0 left, I get a loading indicator.

I'm not sure how to reproduce this

**Comment by tlrobinson** [comment]
> But I wonder if we can make it so that the search term is cleared out after selecting an item, but that the search results persist until you enter a new search term.

One problem with this in the "list" case is there's no way to get back to the full list except by typing a character then deleting it.

**Comment by mazameli** [comment]
> One problem with this in the "list" case is there's no way to get back to the full list except by typing a character then deleting it.

Well, what if when you select an item, the values input box loses focus; then if the values box gains focus again (i.e. you click into the box), the search results get cleared out? I.e., by clicking into the box the user is showing intent to start a new search.

**Comment by mazameli** [comment]
This is a little obscure and edge-casey, but thought I'd at least note it: I pasted an ID value into a dashboard ID filter box, and the value I pasted happened to have trailing whitespace, so it didn't match any results. (I don't know if we can do anything about that since whitespace is sometimes significant.) Then when I deleted the white space and selected the corresponding ID from the list, the selected ID still remained present in the list of options.

weirdness

**Comment by mazameli** [comment]
Looks like the logic for placeholder text when searching through string-based primary or foreign key values/IDs is a little wonky still:

Primary key:

screen shot 2018-02-16 at 4 54 17 pm

Foreign key:

screen shot 2018-02-16 at 4 54 27 pm

File: frontend/src/metabase/components/FieldValuesWidget.jsx
Comment (kdoh) [inline frontend/src/metabase/components/FieldValuesWidget.jsx]: What's the main reason for allowing style as props here? If we can find a way to control specific instance needs without this I'd prefer that rather than allowing arbitrary style.

**Review by kdoh — APPROVED** [review]
Code looks pretty good overall. Couple small points I brought up that we should look at but I wouldn't block merging over.

**Comment by mazameli** [comment]
This is all working well enough now that I think we should merge it. Great work you guys! 👏 

I've created a follow-up polish issue with some things I noticed here: https://github.com/metabase/metabase/issues/6995

**Comment by salsakran** [comment]
I'm good to merge from the UX perspective if in #6995
`The operators being hidden under Options was only intended to be the behavior for categoric fields, not numeric ones. The filter popover for numeric fields should be unchanged.` 
is fixed before the next release. 

@tlrobinson feel free to merge once tests pass.

**Pull request merged**
